// ngp-image-picker/esm2020/lib/ngp-image-picker.module.mjs
import { NgpImagePickerComponent } from './ngp-image-picker.component';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { TabsComponent } from './components/tabs/tabs.component';
import { EditImageComponent } from './components/edit-image/edit-image.component';
import { BasicFiltersComponent } from './components/basic-filters/basic-filters.component';
import { CropperWrapperComponent } from './components/cropper-wrapper/cropper-wrapper.component';
import * as i0 from "@angular/core";
export class NgpImagePickerModule {
}
NgpImagePickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgpImagePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgpImagePickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.4", ngImport: i0, type: NgpImagePickerModule, declarations: [NgpImagePickerComponent, TabsComponent, EditImageComponent, BasicFiltersComponent, CropperWrapperComponent], imports: [CommonModule, FormsModule, ReactiveFormsModule], exports: [NgpImagePickerComponent] });
NgpImagePickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgpImagePickerModule, imports: [CommonModule, FormsModule, ReactiveFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgpImagePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgpImagePickerComponent, TabsComponent, EditImageComponent, BasicFiltersComponent, CropperWrapperComponent],
                    imports: [CommonModule, FormsModule, ReactiveFormsModule],
                    exports: [NgpImagePickerComponent],
                    providers: [],
                }]
        }] });
