// ngp-image-picker/esm2020/lib/components/cropper-wrapper/cropper-wrapper.component.mjs
import { Component, Input, Output, EventEmitter, HostListener } from '@angular/core';
import Croppr from '../../functions/croppr/index';
import * as i0 from "@angular/core";
export class CropperWrapperComponent {
    constructor() {
        this.imageSrc = '';
        this.croppr = null;
        this.croppSize = { width: 150, height: 150 };
        this.croppUpdate = new EventEmitter();
    }
    onKeyDown() {
        if (!this.croppr)
            return;
        this.croppr.options.aspectRatio = 1.0;
    }
    onKeyUp() {
        if (!this.croppr)
            return;
        this.croppr.options.aspectRatio = null;
    }
    set _imageSrc(value) {
        this.imageSrc = value;
    }
    set setSize(value) {
        this.croppSize = value;
        if (this.croppr)
            this.croppr.resizeTo(this.croppSize.width, this.croppSize.height);
    }
    ngOnInit() { }
    ngAfterViewInit() {
        this.croppr = new Croppr('#croppr', {
            minSize: [32, 32, 'px'],
            startSize: [this.croppSize.width, this.croppSize.height, 'px'],
            onInitialize: (data) => {
                this.croppUpdate.emit(data.getValue());
            },
            onCropEnd: (data) => {
                this.croppUpdate.emit(data);
            },
        });
    }
    ngOnDestroy() {
        if (!this.croppr)
            return;
        this.croppr.destroy();
    }
}
CropperWrapperComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: CropperWrapperComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
CropperWrapperComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.0.4", type: CropperWrapperComponent, selector: "lib-cropper-wrapper", inputs: { _imageSrc: "_imageSrc", setSize: "setSize" }, outputs: { croppUpdate: "croppUpdate" }, host: { listeners: { "document:keydown.Control": "onKeyDown($event)", "document:keyup.Control": "onKeyUp($event)" } }, ngImport: i0, template: "<img id=\"image-full\" [src]=\"this.imageSrc\" id=\"croppr\" />\n", styles: [""] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: CropperWrapperComponent, decorators: [{
            type: Component,
            args: [{ selector: 'lib-cropper-wrapper', template: "<img id=\"image-full\" [src]=\"this.imageSrc\" id=\"croppr\" />\n" }]
        }], ctorParameters: function () { return []; }, propDecorators: { onKeyDown: [{
                type: HostListener,
                args: ['document:keydown.Control', ['$event']]
            }], onKeyUp: [{
                type: HostListener,
                args: ['document:keyup.Control', ['$event']]
            }], _imageSrc: [{
                type: Input
            }], setSize: [{
                type: Input
            }], croppUpdate: [{
                type: Output
            }] } });
