// ngp-image-picker/esm2020/lib/functions/croppr/handle.mjs
/**
 * Handle component
 */
export default class Handle {
    constructor(position, constraints, cursor, eventBus) {
        var self = this;
        this.position = position;
        this.constraints = constraints;
        this.cursor = cursor;
        this.eventBus = eventBus;
        // Create DOM element
        this.el = document.createElement("div");
        this.el.className = "croppr-handle";
        this.el.style.cursor = cursor;
        // Attach initial listener
        this.el.addEventListener("mousedown", onMouseDown);
        function onMouseDown(e) {
            e.stopPropagation();
            document.addEventListener("mouseup", onMouseUp);
            document.addEventListener("mousemove", onMouseMove);
            // Notify parent
            self.eventBus.dispatchEvent(new CustomEvent("handlestart", {
                detail: { handle: self },
            }));
        }
        function onMouseUp(e) {
            e.stopPropagation();
            document.removeEventListener("mouseup", onMouseUp);
            document.removeEventListener("mousemove", onMouseMove);
            // Notify parent
            self.eventBus.dispatchEvent(new CustomEvent("handleend", {
                detail: { handle: self },
            }));
        }
        function onMouseMove(e) {
            e.stopPropagation();
            // Notify parent
            self.eventBus.dispatchEvent(new CustomEvent("handlemove", {
                detail: { mouseX: e.clientX, mouseY: e.clientY },
            }));
        }
    }
}
